File: app-sdk/AttributionSDK.kt
/**
 * OneKey Attribution Android SDK
 *
 * 接入方式：
 * 1. 将此文件加入项目
 * 2. 在 App 首次启动时调用 OnekeyAttribution.trackInstall()
 * 3. 在用户注册成功时调用 OnekeyAttribution.trackRegister()
 */

package com.onekey.attribution

import android.content.Context
import android.content.res.Resources
import android.os.Build
import android.webkit.WebSettings
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.Locale
import java.util.TimeZone
import kotlin.concurrent.thread

object OnekeyAttribution {

    // TODO: 替换为你的归因服务地址
    var server = "https://your-attribution-server.com"

    private const val PREFS_NAME = "onekey_attribution"
    private const val KEY_TRACKED = "tracked"
    private const val KEY_CHANNEL = "channel"
    private const val KEY_CAMPAIGN = "campaign"

    /**
     * 事件3：首次打开App
     * 在 Application.onCreate 或首次进入主界面时调用
     */
    fun trackInstall(
        context: Context,
        instanceId: String,
        callback: ((matched: Boolean, channel: String?, campaign: String?) -> Unit)? = null
    ) {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        if (prefs.getBoolean(KEY_TRACKED, false)) {
            val ch = prefs.getString(KEY_CHANNEL, null)
            val ca = prefs.getString(KEY_CAMPAIGN, null)
            callback?.invoke(ch != null, ch, ca)
            return
        }

        val fingerprint = collectFingerprint(context, instanceId)

        thread {
            try {
                val result = post("/api/track/install", fingerprint)
                val match = result?.optJSONObject("match")
                val matched = match?.optBoolean("matched") ?: false
                val channel = match?.optString("channel", null)
                val campaign = match?.optString("campaign", null)

                prefs.edit().apply {
                    putBoolean(KEY_TRACKED, true)
                    if (matched) {
                        putString(KEY_CHANNEL, channel)
                        putString(KEY_CAMPAIGN, campaign)
                    }
                    apply()
                }

                callback?.invoke(matched, channel, campaign)
            } catch (e: Exception) {
                e.printStackTrace()
                callback?.invoke(false, null, null)
            }
        }
    }

    /**
     * 事件4：注册
     */
    fun trackRegister(context: Context, instanceId: String) {
        val body = JSONObject().apply {
            put("instance_id", instanceId)
        }
        thread {
            try { post("/api/track/register", body) } catch (_: Exception) {}
        }
    }

    /**
     * 获取已保存的渠道
     */
    fun getChannel(context: Context): String? {
        return context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            .getString(KEY_CHANNEL, null)
    }

    // ---------- 内部方法 ----------

    private fun collectFingerprint(context: Context, instanceId: String): JSONObject {
        val dm = Resources.getSystem().displayMetrics
        return JSONObject().apply {
            put("screen", "${dm.widthPixels}x${dm.heightPixels}")
            put("language", Locale.getDefault().language)
            put("timezone", TimeZone.getDefault().id)
            put("os", "Android")
            put("os_version", Build.VERSION.RELEASE)
            put("model", Build.MODEL)
            put("app_version", try {
                context.packageManager.getPackageInfo(context.packageName, 0).versionName
            } catch (_: Exception) { "" })
            put("platform", "android")
            put("instance_id", instanceId)
        }
    }

    private fun post(endpoint: String, body: JSONObject): JSONObject? {
        val conn = URL(server + endpoint).openConnection() as HttpURLConnection
        conn.requestMethod = "POST"
        conn.setRequestProperty("Content-Type", "application/json")
        conn.connectTimeout = 10000
        conn.doOutput = true
        conn.outputStream.write(body.toString().toByteArray())
        val response = conn.inputStream.bufferedReader().readText()
        conn.disconnect()
        return JSONObject(response)
    }
}

File: app-sdk/AttributionSDK.swift
/**
 * OneKey Attribution iOS SDK
 *
 * 接入方式：
 * 1. 将此文件加入项目
 * 2. 在 App 首次启动时调用 OnekeyAttribution.trackInstall()
 * 3. 在用户注册成功时调用 OnekeyAttribution.trackRegister()
 */

import UIKit

class OnekeyAttribution {

    // TODO: 替换为你的归因服务地址
    static let server = "https://your-attribution-server.com"

    private static let trackedKey = "onekey_attribution_tracked"
    private static let channelKey = "onekey_attribution_channel"
    private static let campaignKey = "onekey_attribution_campaign"
    private static let paramsKey = "onekey_attribution_params"

    // MARK: - 事件3：首次打开App

    /// 在 App 首次启动时调用（建议在 AppDelegate 或 onboarding 流程开始前）
    /// 只会执行一次，重复调用会被忽略
    static func trackInstall(instanceId: String, completion: ((Bool, String?, String?) -> Void)? = nil) {
        guard !UserDefaults.standard.bool(forKey: trackedKey) else {
            let ch = UserDefaults.standard.string(forKey: channelKey)
            let ca = UserDefaults.standard.string(forKey: campaignKey)
            completion?(ch != nil, ch, ca)
            return
        }

        let fingerprint = collectFingerprint(instanceId: instanceId)

        post(endpoint: "/api/track/install", body: fingerprint) { data in
            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let match = json["match"] as? [String: Any] else {
                completion?(false, nil, nil)
                return
            }

            let matched = match["matched"] as? Bool ?? false
            let channel = match["channel"] as? String
            let campaign = match["campaign"] as? String
            let score = match["score"] as? Double ?? 0

            // 持久化结果
            UserDefaults.standard.set(true, forKey: trackedKey)
            if matched {
                UserDefaults.standard.set(channel, forKey: channelKey)
                UserDefaults.standard.set(campaign, forKey: campaignKey)
                if let params = match["custom_params"] as? [String: Any],
                   let paramsData = try? JSONSerialization.data(withJSONObject: params) {
                    UserDefaults.standard.set(paramsData, forKey: paramsKey)
                }
            }

            print("[Attribution] matched=\(matched), score=\(score), channel=\(channel ?? "none"), campaign=\(campaign ?? "none")")
            completion?(matched, channel, campaign)
        }
    }

    // MARK: - 事件4：注册

    /// 在用户注册成功后调用
    static func trackRegister(instanceId: String) {
        let body: [String: Any] = [
            "instance_id": instanceId,
        ]
        post(endpoint: "/api/track/register", body: body, completion: nil)
    }

    // MARK: - 读取归因结果

    /// 获取已保存的渠道信息（匹配成功后可用）
    static var channel: String? {
        UserDefaults.standard.string(forKey: channelKey)
    }

    static var campaign: String? {
        UserDefaults.standard.string(forKey: campaignKey)
    }

    // MARK: - 内部方法

    private static func collectFingerprint(instanceId: String) -> [String: Any] {
        let screen = UIScreen.main.nativeBounds
        return [
            "screen": "\(Int(screen.width))x\(Int(screen.height))",
            "language": Locale.current.language.languageCode?.identifier ?? "",
            "timezone": TimeZone.current.identifier,
            "os": "iOS",
            "os_version": UIDevice.current.systemVersion,
            "model": UIDevice.current.model,
            "app_version": Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "",
            "platform": "ios",
            "instance_id": instanceId,
        ]
    }

    private static func post(endpoint: String, body: [String: Any], completion: ((Data?) -> Void)?) {
        guard let url = URL(string: server + endpoint) else { return }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.timeoutInterval = 10
        req.httpBody = try? JSONSerialization.data(withJSONObject: body)

        URLSession.shared.dataTask(with: req) { data, _, error in
            if let error = error {
                print("[Attribution] Error: \(error.localizedDescription)")
            }
            completion?(data)
        }.resume()
    }
}
